093.mediawiki

```diff
@@ -0,0 +1,329 @@
+ |header||Block header||80 bytes||The header of the block being provided, encoded as per the header portion of the BLOCK message (NOT including the nTx field of the HEADERS message)
+ |-
+ |vAdditionalTxs||<nowiki>vector<CTransaction></nowiki>||variable||Transactions that the receiver is probably missing. Standard transaction Bitcoin serialization is used.
+ |-
+ |nBlockTxs||uint64_t||8 bytes||Number of Transactions in the block
+ |-
+ |grapheneSet||CGrapheneSet||variable||Encapsulates Graphene set reconciliation logic
+ |}
+
+The vAdditionalTxs field should include at least the coinbase transaction, since it is not possible for it to be in the receiver's mempool.
+
+====CGrapheneSet====
+''CGrapheneSet'' contains all information critical to the Graphene set reconciliation process, absent any block-specific details. It is the only non-standard data structure in ''CGrapheneBlock''.
+
+{|
+ |'''Field Name'''||'''Type'''||'''Size'''||'''Purpose'''
+ |-
+ |ordered||uint8_t||1 byte||1 if order is important in the set and 0 otherwise; other values are reserved for future use
+ |-
+ |nReceiverUniverseItems||uint64_t||8 bytes||The number of transactions in the receiver's mempool
+ |-
+ |encodedRank||<nowiki>vector<unsigned char></nowiki>||ordered * ceil(n log(n))/8 bytes||Order information for items in set
+ |-
+ |setFilter||CBloomFilter||variable||Bloom filter containing items in set
+ |-
+ |setIblt||CIblt||variable||IBLT containing items in set
+ |}
+
+====CBloomFilter====
+''CBloomFilter'' is Bitcoin's standard Bloom filter implementation with one adjustment, described below. It is used as part of the set reconciliation process in ''CGrapheneSet'' to filter out transactions from ''T'' that do not belong to the block.
+
+{|
+ |'''Field Name'''||'''Type'''||'''Size'''||'''Purpose'''
+ |-
+ |vData||<nowiki>vector<unsigned char></nowiki>||variable||Bit array for filter
+ |-
+ |isFull||bool||1 byte||True if every bit in vData is set to 1
+ |-
+ |isEmpty||bool||1 byte||True if every bit in vData is set to 0
+ |-
+ |nHashFuncs||unsigned int||2, 4 bytes||Number of hash functions used in filter
+ |-
+ |nTweak||unsigned int||2, 4, bytes||Additive offset for hash function inputs
+ |-
+ |nFlags||unsigned char||1 byte||Defines behavior for transaction insertion
+ |}
+
+The standard Bloom filter implementation does not always honor the false positive rate (FPR) requested during initialization. This behavior is problematic for Graphene because the FPR determines the number of erroneous transactions removed from IBLT ''I'', which affects the decode rate for ''I''. For this reason, it is required that ''nDesiredSize'' and ''nHashFuncs'' be defined slightly differently than in the standard ''CBloomFilter'' implementation, as follows.
+
+<code>
+nDesiredSize = (unsigned int)(ceil(-1 / LN2SQUARED * nElements * log(nFPRate) / 8));
+</code>
+<br />
+<code>
+nHashFuncs = (unsigned int)max(MIN_N_HASH_FUNC, int(vData.size() * 8 / nElements * LN2));
+</code>
+
+====CIblt====
```

Comment:
I recommend renaming this class to CInverseBloomLookupTable.  Modern IDEs with autocomplete make class name length irrelevant, and reading/writing CIblt becomes a nightmare in some fonts.

Reply:
That sounds reasonable to me. How about C**Invertible**BloomLookupTable instead?

Reply:
Isn't naming of classes something implementations can decide for themselves?

Reply:
Yes, it's just difficult to read in this spec regardless of who's implementing it. ;)